### PR TITLE
Telemetry 15.3

### PR DESCRIPTION
**Pull request by paulvanbrenk**
Add a bunch of telemetry to the NodeTools

### CODE REVIEW

File: Nodejs/Product/Nodejs/Debugger/NodeDebugProvider.cs [hunk 1]
```diff
@@ -117,7 +120,7 @@ private void SetupDebugTargetInfoForWebkitV2Protocol(ref VsDebugTargetInfo vsDeb
                 new JProperty("runtimeExecutable", nodeExe),
                 new JProperty("cwd", cwd),
                 new JProperty("console", "externalTerminal"),
-                new JProperty("diagnosticLogging", NodejsProjectLauncher.CheckEnableDiagnosticLoggingOption()),
+                new JProperty("trace", NodejsProjectLauncher.CheckEnableDiagnosticLoggingOption()),
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Debugger/NodeDebugProvider.cs]: This was a bug I missed porting from the ProjectLauncher, i.e. the non AnyCode debugger.

File: Nodejs/Product/Nodejs/Telemetry/TelemetryHelper.cs [hunk 2]
```diff
@@ -20,8 +23,17 @@ public static void Initialize()
 
         public static void LogProjectImported()
         {
-            defaultSession.PostUserTask(TelemetryEvents.ProjectImported, TelemetryResult.Success);
+            defaultSession.PostUserTask(ProjectImported, TelemetryResult.Success);
+        }
+
+        public static void LogDebuggingStarted(string debuggerName, string nodeVersion, bool isProject = true)
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Telemetry/TelemetryHelper.cs]: Should version be a string??

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs [hunk 3]
```diff
@@ -302,6 +303,9 @@ internal void Install(PackageCatalogEntryViewModel package)
             if (!string.IsNullOrEmpty(package.Name))
             {
                 var selectedVersion = this.SelectedVersion is SemverVersion ? ((SemverVersion)this.SelectedVersion).ToString() : string.Empty;
+
+                TelemetryHelper.LogInstallNpmPackage(package.Name, selectedVersion);
```
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: @bowdenk7 Can we actually log this, or is this PII?

File: Nodejs/Product/TestAdapter/TestExecutor.cs [hunk 4]
```diff
@@ -182,6 +183,16 @@ public void RunTests(IEnumerable<TestCase> tests, IRunContext runContext, IFrame
             }
         }
 
+        private void LogTelemetry(int testCount, bool isDebugging)
+        {
+            var userTask = new UserTaskEvent("VS/NodejsTools/UnitTestsExecuted", TelemetryResult.Success);
+            userTask.Properties["VS.NodejsTools.TestCount"] = testCount;
+            userTask.Properties["VS.NodejsTools.IsDebugging"] = isDebugging;
```
Comment (paulvanbrenk) [inline Nodejs/Product/TestAdapter/TestExecutor.cs]: note: debugger version is more usefull

**Comment by amcasey** [comment]
LGTM, modulo the comments @paulvanbrenk added on my behalf.

File: Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs (hunk 3 above)
Comment (billti) [inline Nodejs/Product/Nodejs/NpmUI/NpmPackageInstallViewModel.cs]: We shouldn't log this. They could be using a private registry.

File: Nodejs/Product/Nodejs/Nodejs.csproj [hunk 5]
```diff
@@ -1319,4 +1318,5 @@
   <Target Name="_DeleteAfterCreatePkgDef" AfterTargets="GeneratePkgDef">
     <Delete Files="@(_CopiedForCreatePkgdef)" />
   </Target>
+  <Import Project="..\..\packages\Microsoft.Diagnostics.Tracing.EventSource.Redist.1.1.16-beta\build\portable-net45+win8+wpa81\Microsoft.Diagnostics.Tracing.EventSource.Redist.targets" Condition="Exists('..\..\packages\Microsoft.Diagnostics.Tracing.EventSource.Redist.1.1.16-beta\build\portable-net45+win8+wpa81\Microsoft.Diagnostics.Tracing.EventSource.Redist.targets')" />
```
Comment (billti) [inline Nodejs/Product/Nodejs/Nodejs.csproj]: Do we require this version over the standard System.Diagnostics.Tracing.EventSource class already in the .NET FX?

File: Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs [hunk 6]
```diff
@@ -103,6 +104,8 @@ private void Run()
                 return Enumerable.Empty<IPackage>();
             }
 
+            TelemetryHelper.LogSearchNpm(filterText);
```
Comment (billti) [inline Nodejs/Product/Nodejs/NpmUI/NpmWorker.cs]: Likewise, logging what the user has typing in the search box may have PII issues.

File: Nodejs/Product/Nodejs/Project/NodejsProjectLauncher.cs [hunk 7]
```diff
@@ -71,27 +72,32 @@ private int Start(string file, bool debug)
...
 
-            var chromeProtocolRequired = Nodejs.GetNodeVersion(nodePath) >= new Version(8, 0) || CheckDebugProtocolOption();
+            var nodeVersion = Nodejs.GetNodeVersion(nodePath);
+            var chromeProtocolRequired = nodeVersion >= new Version(8, 0) || CheckDebugProtocolOption();
             var startBrowser = ShouldStartBrowser();
 
             if (debug && !chromeProtocolRequired)
             {
                 StartWithDebugger(file);
+                TelemetryHelper.LogDebuggingStarted("Node6", nodeVersion.ToString());
```
Comment (billti) [inline Nodejs/Product/Nodejs/Project/NodejsProjectLauncher.cs]: Does `Node6` here really mean < 8?

File: Nodejs/Product/Nodejs/Telemetry/TelemetryHelper.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            userTask.Properties[NodeVersion] = nodeVersion;
+            userTask.Properties[IsProject] = isProject;
+
+            defaultSession.PostEvent(userTask);
+        }
+
+        public static void LogSearchNpm(string query)
+        {
+            var userTask = new UserTaskEvent(SearchNpm, TelemetryResult.Success);
+            userTask.Properties[NpmSearchQuery] = query;
```
Comment (billti) [inline Nodejs/Product/Nodejs/Telemetry/TelemetryHelper.cs]: Likewise with the PII

File: Nodejs/Product/Nodejs/packages.config [hunk 9]
```diff
@@ -3,7 +3,11 @@
   <package id="MicroBuild.Core" version="0.2.0" targetFramework="net46" />
   <package id="Microsoft.ApplicationInsights" version="2.2.0" targetFramework="net46" />
   <package id="Microsoft.ApplicationInsights.PersistenceChannel" version="1.2.3" targetFramework="net46" />
+  <package id="Microsoft.Diagnostics.Tracing.EventSource.Redist" version="1.1.16-beta" targetFramework="net46" />
+  <package id="Microsoft.VisualStudio.RemoteControl" version="14.0.262-masterA5CACE98" targetFramework="net46" />
```
Comment (billti) [inline Nodejs/Product/Nodejs/packages.config]: What is this `RemoteControl` package for?

File: Nodejs/Product/Nodejs/Nodejs.csproj (hunk 5 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Nodejs.csproj]: This came with the telemetry NuGet package, I didn't want to mess with that.

File: Nodejs/Product/Nodejs/Project/NodejsProjectLauncher.cs (hunk 7 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/Project/NodejsProjectLauncher.cs]: this means the node6 protocol..

File: Nodejs/Product/Nodejs/packages.config (hunk 9 above)
Comment (paulvanbrenk) [inline Nodejs/Product/Nodejs/packages.config]: This came with the telemetry NuGet package, I didn't want to mess with that.

**Comment by paulvanbrenk** [comment]
Pushed this in, so we would make the cutoff for the VS build.

**Pull request merged**